discord/ext/commands/core.py

```diff
@@ -476,31 +476,42 @@ async def _actual_conversion(self, ctx, converter, argument, param):
             raise BadArgument('Converting to "{}" failed for parameter "{}".'.format(name, param.name)) from exc
 
     async def do_conversion(self, ctx, converter, argument, param):
-        try:
-            origin = converter.__origin__
-        except AttributeError:
-            pass
-        else:
-            if origin is typing.Union:
-                errors = []
-                _NoneType = type(None)
-                for conv in converter.__args__:
-                    # if we got to this part in the code, then the previous conversions have failed
-                    # so we should just undo the view, return the default, and allow parsing to continue
-                    # with the other parameters
-                    if conv is _NoneType and param.kind != param.VAR_POSITIONAL:
-                        ctx.view.undo()
-                        return None if param.default is param.empty else param.default
-
-                    try:
-                        value = await self._actual_conversion(ctx, conv, argument, param)
-                    except CommandError as exc:
-                        errors.append(exc)
-                    else:
+        origin = typing.get_origin(converter)
+
+        if origin is typing.Union:
+            errors = []
+            _NoneType = type(None)
+            for conv in typing.get_args(converter):
+                # if we got to this part in the code, then the previous conversions have failed
+                # so we should just undo the view, return the default, and allow parsing to continue
+                # with the other parameters
+                if conv is _NoneType and param.kind != param.VAR_POSITIONAL:
+                    ctx.view.undo()
+                    return None if param.default is param.empty else param.default
+
+                try:
+                    value = await self.do_conversion(ctx, conv, argument, param)
+                except CommandError as exc:
+                    errors.append(exc)
+                else:
+                    return value
+
+            # if we're here, then we failed all the converters
+            raise BadUnionArgument(param, typing.get_args(converter), errors)
+
+        if origin is typing.Literal:
+            errors = []
+            for literal in self._flattened_typing_literal_args(converter):
+                try:
+                    value = await self._actual_conversion(ctx, type(literal), argument, param)
+                except CommandError as exc:
+                    errors.append(exc)
+                else:
+                    if value == literal:
                         return value
 
-                # if we're  here, then we failed all the converters
-                raise BadUnionArgument(param, converter.__args__, errors)
+            # if we're here, then we failed to match all the literals
+            raise BadLiteralArgument(param, tuple(self._flattened_typing_literal_args(converter)), errors)
```

Comment:
Use a shared variable from the above for loop for the flattened args.

Reply:
`self._flattened_typing_literal_args` is currently a generator, so it can't be reused in this way. Unless you think this should be changed?

Reply:
I suggest storing the output of the generator in a tuple before iterating at the start.
And iterating over that tuple as well as passing it into this exception's constructor, saves the function call.

Reply:
Alright, cool. Would there possibly be some performance penalty to pre-generating the generator for the for-loop since it can exit early? When the exception isn't raised.

Reply:
likely yes, but i wouldn't be able to tell you if running the generator twice is better or not.

could keep a list of args and append to it as you iterate, as you'd only ever need the full list for the exception if it had already exhausted all args.

Reply:
Appends might become slow, at least if there's a lot of literals. The penalty is probably negligible either way. Pre-generation seems the simplest and looks the best, I think, so I'll just go ahead with that solution.